arches/app/models/models.py

```diff
@@ -2106,3 +2115,56 @@ def __str__(self):
     class Meta:
         managed = True
         db_table = "spatial_views"
+
+    def save(self, *args, **kwargs):
```

Comment:
@jacobtylerwalls pointed out to me last week that the `clean` method is provided by Django for validation. It would probably be better to use that here rather than overriding the default save method.

Reply:
@chiatt @jacobtylerwalls - `clean()` is not called on model.save() so if spatialview model is accessed outside the API, we would need to hope it calls clean first.  

_Note, however, that like [Model.full_clean()](https://docs.djangoproject.com/en/4.2/ref/models/instances/#django.db.models.Model.full_clean), a model’s clean() method is not invoked when you call your model’s [save()](https://docs.djangoproject.com/en/4.2/ref/models/instances/#django.db.models.Model.save) method._

https://docs.djangoproject.com/en/4.2/ref/models/instances/#django.db.models.Model.clean

The alternative is that I could try using use a pre_save signal to invoke instance.clean?

Reply:
> if spatialview model is accessed outside the API, we would need to hope it calls clean first.

I think this is reasonable to expect. Django firmly treats validation and persistence as two separate concerns, so we can/should expect applications to be using the framework as designed. If one relies on the database to reject invalid input, the following goodies go missing, and IMO we shouldn't continue to encourage the practice:
- model field validation e.g. URLField validating schemes, PositiveIntegerField validating sign, any argument passed to validators= (save()/postgres doesn't know about any of this)
- blank=False (save/postgres doesn't know about this)
- models/fields that did implement their own clean()

Then, it also becomes harder for applications (like the controlled list manager for example) trying to do the more idiomatic thing of calling full_clean() before save():
- they will have to know to catch ValidationError from a save() call, which is unusual, and wrap it the same way they wrap errors from clean(), which could be error-prone
- full_clean() will now return a misleading result

(On this last point, are there subtle breakages in the Django admin if full_clean returns a false positive success?)

> The alternative is that I could try using use a pre_save signal to invoke instance.clean?

I understand the impulse to try to make this validation "nonoptional" somehow, but I think this would reproduce the same problem: I would be surprised to learn that my save() calls might emit ValidationError, when I would expect this only from clean(). I would still need to wrap my save() calls with try, and I will still get misleading success from full_clean().

In my view, we need to get comfortable with the idea that there are various ways to get around validation, including signals, bulk_create/bulk_update, and QuerySet.update(), and not add code that both incompletely plugs the holes and also suggests that validation and persistence belong together in Django.

Overriding save() also has other vectors for bugs if you're setting model fields, which thankfully is not happening here. But it does suggest to me that overriding save should be saved (lol) for overriding the specifics of the save mechanism: database routing, forcing inserts v. updates, that kind of thing.

The last line of defense is obviously the database, so if some of these checks could be expressed as database constraints, you could have your cake and eat it too, since full_clean() now checks constraints. (But some of these checks check multiple tables, so likely not.)

Reply:
@jacobtylerwalls - OK, I'm happy to move the functionality to `clean` - sounds like it is going to be more widely present anyway in the codebase. I guess if they don't call `clean` they will just run into database exceptions!

Reply:
done